File: server/schemas.py
# from marshmallow import Schema, fields, validate

# # Tenant schema for input validation and serialization
# class TenantSchema(Schema):
#     id = fields.Int(dump_only=True)
#     tenant_name = fields.String(required=True, validate=validate.Length(min=1, max=100))
#     tenant_phone_number = fields.String(required=True, validate=validate.Length(min=10, max=10))  # Assuming 10-digit numbers
#     house_number = fields.String(required=True, validate=validate.Length(min=1, max=10))
#     house_type = fields.String(required=True, validate=validate.Length(min=1, max=50))
#     deposit_paid = fields.Float(required=True)
#     payment_date = fields.DateTime(dump_only=True)
#     receipt_number_deposit = fields.String(dump_only=True)  # Generated, hence dump_only
#     rent_amount = fields.Float(required=True)
#     due_date = fields.DateTime(dump_only=True)  # Automatically set, hence dump_only
#     rent_receipt_number = fields.String(dump_only=True)
#     rent_paid = fields.Float()
#     amount_due = fields.Float(required=True)

#     # Relationship to Property
#     property_id = fields.Int(required=True)
#     property = fields.Nested('PropertySchema', only=("property_name", "location"), dump_only=True)  # Display basic info only

# # Landlord schema
# class LandlordSchema(Schema):
#     id = fields.Int(dump_only=True)
#     landlord_name = fields.String(required=True, validate=validate.Length(min=1, max=100))
#     phone_number = fields.String(required=True, validate=validate.Length(min=10, max=20))
#     email = fields.String(required=True, validate=validate.Email())

#     # Relationship to Properties
#     properties_owned = fields.Nested('PropertySchema', many=True, only=("id", "property_name"), dump_only=True)

# # Property schema
# class PropertySchema(Schema):
#     id = fields.Int(dump_only=True)  # ID should remain dump_only
#     property_name = fields.String(required=True, validate=validate.Length(min=1, max=100))
#     location = fields.String(required=True, validate=validate.Length(min=1, max=100))
#     landlord_id = fields.Int(required=True)
#     number_of_rooms = fields.Int(required=True)
#     is_occupied = fields.Boolean()  # Remove dump_only to allow updates
#     house_number = fields.String(required=True, validate=validate.Length(min=1, max=10))
#     occupied_rooms = fields.Int()  # Remove dump_only to allow updates
#     price_bedsitter = fields.Float()
#     price_one_bedroom = fields.Float()
#     price_two_bedroom = fields.Float()
#     # Relationship to Tenants
#     tenants = fields.Nested('TenantSchema', many=True, only=("id", "tenant_name", "house_number"), dump_only=True)

#     # Relationship to Landlord
#     landlord = fields.Nested('LandlordSchema', only=("id", "landlord_name"), dump_only=True)
# # User schema for admin and user data
# class UserSchema(Schema):
#     id = fields.Int(dump_only=True)
#     username = fields.String(required=True, validate=validate.Length(min=1, max=100))
#     role = fields.String(validate=validate.OneOf(["admin", "user"]))
#     created_at = fields.DateTime(dump_only=True)

# from marshmallow import Schema, fields, validate

# # Tenant schema for input validation and serialization
# class TenantSchema(Schema):
#     id = fields.Int(dump_only=True)
#     tenant_name = fields.String(required=True, validate=validate.Length(min=1, max=100))
#     tenant_phone_number = fields.String(required=True, validate=validate.Length(equal=10))  # 10-digit numbers
#     house_number = fields.String(required=True, validate=validate.Length(min=1, max=10))
#     house_type = fields.String(required=True, validate=validate.Length(min=1, max=50))
#     deposit_paid = fields.Float(required=True)
#     payment_date = fields.DateTime(dump_only=True)
#     receipt_number_deposit = fields.String(dump_only=True)  # Generated, hence dump_only
#     rent_amount = fields.Float(required=True)
#     due_date = fields.DateTime(dump_only=True)  # Automatically set, hence dump_only
#     rent_receipt_number = fields.String(dump_only=True)
#     rent_paid = fields.Float()
#     amount_due = fields.Float(required=True)
#     property_id = fields.Int(required=True)

#     # Nested relationship to Property
#     property = fields.Nested('PropertySchema', only=("property_name", "location"), dump_only=True)

# # Landlord schema
# class LandlordSchema(Schema):
#     id = fields.Int(dump_only=True)
#     landlord_name = fields.String(required=True, validate=validate.Length(min=1, max=100))
#     phone_number = fields.String(required=True, validate=validate.Length(min=10, max=20))
#     email = fields.String(required=True, validate=validate.Email())

#     # Nested relationship to Properties
#     properties_owned = fields.Nested('PropertySchema', many=True, only=("id", "property_name"), dump_only=True)

# # Property schema
# class PropertySchema(Schema):
#     id = fields.Int(dump_only=True)
#     property_name = fields.String(required=True, validate=validate.Length(min=1, max=100))
#     location = fields.String(required=True, validate=validate.Length(min=1, max=100))
#     landlord_id = fields.Int(required=True)
#     number_of_rooms = fields.Int(required=True)
#     is_occupied = fields.Boolean()
#     house_number = fields.String(required=True, validate=validate.Length(min=1, max=10))
#     occupied_rooms = fields.Int()
#     price_bedsitter = fields.Float()
#     price_one_bedroom = fields.Float()
#     price_two_bedroom = fields.Float()

#     # Nested relationships
#     tenants = fields.Nested('TenantSchema', many=True, only=("id", "tenant_name", "house_number"), dump_only=True)
#     landlord = fields.Nested('LandlordSchema', only=("id", "landlord_name"), dump_only=True)

# # User schema for admin and user data
# class UserSchema(Schema):
#     id = fields.Int(dump_only=True)
#     username = fields.String(required=True, validate=validate.Length(min=1, max=100))
#     role = fields.String(validate=validate.OneOf(["admin", "user"]))
#     created_at = fields.DateTime(dump_only=True)

from marshmallow import Schema, fields, validate

# Tenant schema for input validation and serialization
class TenantSchema(Schema):
    id = fields.Int(dump_only=True)
    tenant_name = fields.String(validate=validate.Length(min=1, max=100), required=False)  # Optional
    tenant_phone_number = fields.String(validate=validate.Length(equal=10), required=False)  # Optional
    house_number = fields.String(validate=validate.Length(min=1, max=10), required=False)  # Optional
    house_type = fields.String(validate=validate.Length(min=1, max=50), required=False)  # Optional
    deposit_paid = fields.Float(required=False)  # Optional
    payment_date = fields.DateTime(dump_only=True)  # Automatically set, hence dump_only
    receipt_number_deposit = fields.String(dump_only=True)  # Generated, hence dump_only
    rent_amount = fields.Float(required=False)  # Optional
    due_date = fields.DateTime(dump_only=True)  # Automatically set, hence dump_only
    rent_receipt_number = fields.String(dump_only=True)  # Generated, hence dump_only
    rent_paid = fields.Float(required=False)  # Optional
    amount_due = fields.Float(required=True)  # Still required
    property_id = fields.Int(required=True)  # Still required

    # Nested relationship to Property
    property = fields.Nested('PropertySchema', only=("property_name", "location"), dump_only=True)


# Landlord schema
class LandlordSchema(Schema):
    id = fields.Int(dump_only=True)
    landlord_name = fields.String(validate=validate.Length(min=1, max=100), required=False)  # Optional
    phone_number = fields.String(validate=validate.Length(min=10, max=20), required=False)  # Optional
    email = fields.String(validate=validate.Email(), required=False)  # Optional

    # Nested relationship to Properties
    properties_owned = fields.Nested('PropertySchema', many=True, only=("id", "property_name"), dump_only=True)


# Property schema
class PropertySchema(Schema):
    id = fields.Int(dump_only=True)
    property_name = fields.String(validate=validate.Length(min=1, max=100), required=False)  # Optional
    location = fields.String(validate=validate.Length(min=1, max=100), required=False)  # Optional
    landlord_id = fields.Int(required=True)  # Still required
    number_of_rooms = fields.Int(required=False)  # Optional
    is_occupied = fields.Boolean(required=False)  # Optional
    house_number = fields.String(validate=validate.Length(min=1, max=10), required=False)  # Optional
    occupied_rooms = fields.Int(required=False)  # Optional
    price_bedsitter = fields.Float(required=False)  # Optional
    price_one_bedroom = fields.Float(required=False)  # Optional
    price_two_bedroom = fields.Float(required=False)  # Optional

    # Nested relationships
    tenants = fields.Nested('TenantSchema', many=True, only=("id", "tenant_name", "house_number"), dump_only=True)
    landlord = fields.Nested('LandlordSchema', only=("id", "landlord_name"), dump_only=True)


# User schema for admin and user data
class UserSchema(Schema):
    id = fields.Int(dump_only=True)
    username = fields.String(validate=validate.Length(min=1, max=100), required=False)  # Optional
    password_hash = fields.String(dump_only=True)  # Keep it as dump_only
    role = fields.String(validate=validate.OneOf(["admin", "user"]), required=False)  # Optional
    created_at = fields.DateTime(dump_only=True)  # Automatically set, hence dump_only

File: client/src/schemas/Schemas.js
// validationSchemas.js
import * as Yup from 'yup';

// Tenant validation schema
export const TenantSchema = Yup.object().shape({
  tenant_name: Yup.string()
    .required('Tenant name is required')
    .min(1, 'Tenant name must be at least 1 character long')
    .max(100, 'Tenant name cannot exceed 100 characters'),
  tenant_phone_number: Yup.string()
    .required('Phone number is required')
    .length(10, 'Phone number must be exactly 10 digits'),
  house_number: Yup.string()
    .required('House number is required')
    .min(1, 'House number must be at least 1 character long')
    .max(10, 'House number cannot exceed 10 characters'),
  house_type: Yup.string()
    .required('House type is required')
    .min(1, 'House type must be at least 1 character long')
    .max(50, 'House type cannot exceed 50 characters'),
  deposit_paid: Yup.number()
    .required('Deposit paid is required')
    .min(0, 'Deposit must be at least 0'),
  payment_date: Yup.date().nullable(),
  receipt_number_deposit: Yup.string().nullable(),
  rent_amount: Yup.number()
    .required('Rent amount is required')
    .min(0, 'Rent amount must be at least 0'),
  due_date: Yup.date().nullable(),
  rent_receipt_number: Yup.string().nullable(),
  rent_paid: Yup.number().nullable().min(0, 'Rent paid must be at least 0'),
  amount_due: Yup.number()
    .required('Amount due is required')
    .min(0, 'Amount due must be at least 0'),
  property_id: Yup.number()
    .required('Property ID is required')
    .integer('Property ID must be an integer'),
});

// Landlord validation schema
export const LandlordSchema = Yup.object().shape({
  landlord_name: Yup.string()
    .required('Landlord name is required')
    .min(1, 'Landlord name must be at least 1 character long')
    .max(100, 'Landlord name cannot exceed 100 characters'),
  phone_number: Yup.string()
    .required('Phone number is required')
    .min(10, 'Phone number must be at least 10 characters long')
    .max(20, 'Phone number cannot exceed 20 characters'),
  email: Yup.string()
    .required('Email is required')
    .email('Email must be a valid email address'),
});

// Property validation schema
export const PropertySchema = Yup.object().shape({
  property_name: Yup.string()
    .required('Property name is required')
    .min(1, 'Property name must be at least 1 character long')
    .max(100, 'Property name cannot exceed 100 characters'),
  location: Yup.string()
    .required('Location is required')
    .min(1, 'Location must be at least 1 character long')
    .max(100, 'Location cannot exceed 100 characters'),
  landlord_id: Yup.number()
    .required('Landlord ID is required')
    .integer('Landlord ID must be an integer'),
  number_of_rooms: Yup.number()
    .required('Number of rooms is required')
    .integer('Number of rooms must be an integer')
    .min(1, 'Number of rooms must be at least 1'),
  is_occupied: Yup.boolean().nullable(),
  house_number: Yup.string()
    .required('House number is required')
    .min(1, 'House number must be at least 1 character long')
    .max(10, 'House number cannot exceed 10 characters'),
  occupied_rooms: Yup.number().nullable().integer('Occupied rooms must be an integer'),
  price_bedsitter: Yup.number().nullable().min(0, 'Price must be at least 0'),
  price_one_bedroom: Yup.number().nullable().min(0, 'Price must be at least 0'),
  price_two_bedroom: Yup.number().nullable().min(0, 'Price must be at least 0'),
});

// User validation schema
export const UserSchema = Yup.object().shape({
  username: Yup.string()
    .required('Username is required')
    .min(1, 'Username must be at least 1 character long')
    .max(100, 'Username cannot exceed 100 characters'),
  role: Yup.string()
    .required('Role is required')
    .oneOf(['admin', 'user'], 'Role must be either admin or user'),
});
